feat: add recursion protection

ShaunSHamilton opened the pull request:
Currently, this has broken the whileloop, forloop implementation.

Also, it only works for arrow function components:
```jsx
const Items = () => {
  return return <h1>Current Quantity of Items in Cart: {props.quantity} <Items /></h1>
}
```
Compiles to:
```jsx
var _LPC = 1;

var _LP = Date.now();

var Items = function Items(props) {
  if (_LPC++ % 2000 === 0 && Date.now() - _LP > 1500) {
    (function (line) {
      console.log("Potential infinite loop detected on line " + line + ". Tests may be failing because of this.");
    })(1, 0);

    return;
  }

  return /*#__PURE__*/React.createElement("h1", null, "Current Quantity of Items in Cart: ", props.quantity);
};

```

Oliver, I have opened this for you to potentially help with, and give tips, if/when you get time.

_The Babel AST is definitely well-featured for expansion, but, man, is it pushing my limits_

Closes https://github.com/freeCodeCamp/freeCodeCamp/issues/36118

Commit:
feat: add recursion protection for React Components

Commit:
broken: VariableDeclaration implimentation

ShaunSHamilton commented:
How should Class components be transformed for recursion protection?

ShaunSHamilton commented:
As I have it for Class Components:
This code:
```
class ShoppingCart extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    { /* Change code below this line */ }
    return <ShoppingCart />
    { /* Change code above this line */ }
  }
};
```
Gets transpiled into:
```
var _LPC = 1;

var _LP = Date.now();

var ShoppingCart = /*#__PURE__*/function (_React$Component) {
  _inherits(ShoppingCart, _React$Component);

  var _super = _createSuper(ShoppingCart);

  function ShoppingCart(props) {
    _classCallCheck(this, ShoppingCart);

    return _super.call(this, props);
  }

  _createClass(ShoppingCart, [{
    key: "render",
    value: function render() {
      if (_LPC++ % 2000 === 0 && Date.now() - _LP > 1500) {
        (function (line) {
          console.log("Potential infinite loop detected on line " + line + ". Tests may be failing because of this.");
        })(9, 0);

        return;
      }

      {
        /* Change code below this line */
      }
      return /*#__PURE__*/React.createElement(Items, null);
      {
        /* Change code above this line */
      }
    }
  }]);

  return ShoppingCart;
}(React.Component);
```

Now, I need to work on combining, and testing for both. **Then**, the real problem of ensuring I can get the loop protection to work again.

Commit:
add class and functional component RP

ShaunSHamilton commented:
Currently, this code is protected against:
```
const Items = (props) => {
  return <h1>Current Quantity of Items in Cart: {props.quantity} <Items /></h1>
}

Items.defaultProps = {
  quantity: 0
}

class ShoppingCart extends React.Component {
  constructor(props) {
    super(props);
  }
  render() {
    { /* Change code below this line */ }
    return <ShoppingCart />
    { /* Change code above this line */ }
  }
};
```

Commit:
fix: variable and class declarations protections

Commit:
clean up. Mostly works

ojeytonwilliams commented:
Hey Shaun. I didn't get notifications about this, which I wasn't expecting given that it's my fork.

I'll take a look at this as soon as I have time, but you'll have to bear with me as I'm a bit rusty!

ShaunSHamilton commented:
Just some thoughts I want to keep track of:
1) Can this package be updated to ES6+ features? Particularly, I would like to make use of optional chaining.
2) Is there some security-related reason for the current _argument-passing_ pattern? That is, why are the function arguments defined and grabbed into a `_ref` variable, instead of directly?

ojeytonwilliams commented:
> Can this package be updated to ES6+ features? Particularly, I would like to make use of optional chaining.

Sure, makes sense.  Go for it.

> Is there some security-related reason for the current argument-passing pattern? That is, why are the function arguments defined and grabbed into a _ref variable, instead of directly?

Well, they don't use that pattern.  It looks like you've been working on the `dist` files, which have been transpiled by Babel.

ShaunSHamilton commented:
>  It looks like you've been working on the dist files, which have been transpiled by Babel.

I feel quite dumb for that - I have been confusing the way I am testing changes, and the actual package code. Thank you, for pointing it out

Commit:
update deps, add optional-chaining, correct change

ShaunSHamilton commented:
Once I figure out how the testing works, I will add tests for the recurssion

ojeytonwilliams commented:
@ShaunSHamilton don't worry about it. It's just the price of doing open source - all the silly stuff we do is public.

What issues are you running into with testing?

ShaunSHamilton commented:
> What issues are you running into with testing?

Nothing specifically, with the testing. I just thought it best to add some tests to the `.test.js` files, as they have nothing to do with recursion at the moment.

Currently, I am just struggling to test this package as a live package. I have posted the issue on the `chat.freecodecamp` `/software-development` channel. I am missing some knowledge on how to deploy and use packages from GitHub, I think.

Commit:
add shaunshamilton to package.json url

Commit:
increase version

ojeytonwilliams commented:
Okay, sure.  The tests really helped me when I was working on this before.

I don't honestly know why `npm i ShaunSHamilton/loop-protect` doesn't work, but 

```
npm remove @freecodecamp/loop-protect
npm i https://github.com/ShaunSHamilton/loop-protect/tarball/master
cd client/node_modules/@freecodecamp/loop-protect
npm i
```
does.  I'm sure there's a better way, but this at least works.

ShaunSHamilton commented:
This is what I had to do to get this working:
1) `cd client`
2) `npm remove @freecodecamp/loop-protect`
3) `npm i https://github.com/ShaunSHamilton/loop-protect/tarball/master`
4) `cd client/node_modules/@freecodecamp/loop-protect`
5) `npm i`
6) `mkdir dist`
7) `npm run build`
8) Check for `dist > index.js`, if it is not there: remove `--copy-files` arg from `package.json > script > build`
9) Now, you may develop the client.

I am expecting this to be sorted, once this PR is merged.

Commit:
add some tests, fix missing breakstatement

ShaunSHamilton commented:
@ojeytonwilliams This is working how I expect it to. Only issue is the tests are not complete, because I do not know how to add `React` to be used in an `eval()`. Hoping on a call might be really helpful here.

ojeytonwilliams commented:
Sure thing.  Let me investigate a little, so I don't waste your time on the call.

Commit:
add tests for recursion protection

ShaunSHamilton commented:
@ojeytonwilliams This should be ready to go

ojeytonwilliams commented on package.json:
```diff
@@ -1,13 +1,13 @@
   "main": "dist/",
-  "version": "2.2.1",
-  "homepage": "https://github.com/ojeytonwilliams/loop-protect",
+  "version": "3.0.1",
```
```suggestion
  "version": "2.3.0",
```
Nitpicky, I know, but this is backwards compatible isn't it?

ojeytonwilliams commented on package.json:
```diff
@@ -19,25 +19,22 @@
-        }
-      ]
+      "@babel/preset-env",
+      "@babel/preset-react"
```
Can we keep the targets?  

```
"targets": {
  "browsers": [">0.25%", "not dead"]
}
```

For our purposes it should match the fCC/fCC repo.

ShaunSHamilton replied:
To be honest, I just removed them, because I kept getting errors, after adding `preset-react`. So, I am not sure where they need to go.

ojeytonwilliams commented on test/loop-protect.test.js:
```diff
@@ -89,7 +89,7 @@ describe('loop', function() {
 
   it('console error when passing string', () => {
-    const code = `var i = 0; while (true) i++; return true`;
+    const code = `var i = 0; while (true) i++; return true;`;
```
```suggestion
    const code = `var i = 0; while (true) i++; return true`;
```
Not that it matters, but this change seems unnecessary.

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        spy = sinon.spy(run);
+    });
+
+    // https://github.com/jsbin/loop-protect/issues/16
```
```suggestion
```

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        })()`;
+
+        const compiled = loopProtect(code);
+        // assert(run(compiled) === 124);
```
```suggestion
```

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+    it('should rewrite class components', function () {
+        var c = code.simpleclass;
+        var compiled = loopProtect(c);
+        // assert(compiled !== c);
```
```suggestion
        expect(compiled).not.toBe(c);
```
We're checking for rewriting, after all.

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        var compiled = loopProtect(c);
+        // assert(compiled !== c);
+        var result = run(compiled);
+        // assert(result === 1);
```
```suggestion
```

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        var c = code.infiniteclass;
+        var compiled = loopProtect(c);
+
+        // assert(compiled !== c);
+        // assert(spy(compiled) === true);
```
```suggestion
```

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        var c = code.simplefunctionalclass;
+        var compiled = loopProtect(c);
+        var r = run(compiled);
+        // expect(compiled).not.toBe(c);
```
```suggestion
        expect(compiled).not.toBe(c);
```
We do expect it to be modified, right?  The modified function still works, as shown by the next assertion, but it should have been changed.

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+    it('should handle non-keyword "class"', function () {
+        var c = code.notclass;
+        var compiled = loopProtect(c);
+        // assert(compiled == c);
```
```suggestion
```

ojeytonwilliams commented on test/recursion-protect.test.js:
```diff
@@ -0,0 +1,106 @@
+        var compiled = loopProtect(c);
+        // assert(compiled == c);
+        var result = run(compiled);
+        // assert(result === true);
```
```suggestion
```

ojeytonwilliams reviewed (changes requested):
Nice one, Shaun.  I found a few things to nitpick, but this is impressive work.

ShaunSHamilton committed:
remove comments, correct versioning

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

Commit:
update lock, and add browser targets

ShaunSHamilton commented:
Hey Oliver, no rush, but do you have any more thoughts on this?

ojeytonwilliams commented:
Hey, so I tried this out in the main repo, but the tests started to fail.  Specifically because Babel is having problems with `.replace`

I see
```
"unknown: Cannot read property 'unshift' of undefined"
```

when the editor gets Babel to evaluate

```
const removeAssignments = (str) => str.replace(/myNoun\s*=\s*["']dog["']/g, '')
```

in Word Blanks.

Do you see that when you import this into the main repo?

ShaunSHamilton commented:
> Do you see that when you import this into the main repo?

Yes, I see this as well as some issues with the React challenges. I will add tests for these cases, and see if I can fix them.

ojeytonwilliams commented:
Righto. Let me know if I can help.

Commit:
add optional checking and .replace test

Commit:
prevent unshift, add stacked arrow func test

Commit:
add enzyme, broken

ojeytonwilliams committed:
fix: update package-lock

ojeytonwilliams committed:
chore: downgrade react to 16

The react 17 adapter is still in progress, so we should wait for that
before using the latest react version.

Commit:
increase timeout for React tests

ShaunSHamilton commented:
@ojeytonwilliams My suggestion is we merge this, as is.

We can publish a new version, then test with that, as you suggested. Also, once this is merged, we can transfer this repo to `freeCodeCamp`.

ojeytonwilliams commented:
@ShaunSHamilton it's been a while, but this looks okay. Good enough for testing at any rate.
